### PR TITLE
Changes NativeEngineKNNQuery to execute search and rescore in createWeight phase

### PR DESCRIPTION
**Pull request by shatejas**
rewrite is executed in QueryPhase and FetchPhase in opensearch, this increases the query latency. As an alternative, the search logic is moved to createWeight phase. This will only be used when there is a rescore context present, as a result this will only execute when the mode is set to on-disk

Screenshot 2024-08-26 at 9 45 41 AM

### Check List
- [x] New functionality includes testing.
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java [hunk 1]
```diff
@@ -96,7 +95,7 @@ public static Query create(CreateQueryRequest createQueryRequest) {
                         .rescoreContext(rescoreContext)
                         .build();
             }
-            return isKnnQueryRewriteEnabled() ? new NativeEngineKnnVectorQuery(knnQuery) : knnQuery;
+            return createQueryRequest.getRescoreContext().isPresent() ? new NativeEngineKnnVectorQuery(knnQuery) : knnQuery;
```
Comment (shatejas) [inline src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java:98]: @jmazanec15 Please verify this condition is correct

File: CHANGELOG.md [hunk 2]
```diff
@@ -39,8 +39,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 * Refactor method structure and definitions [#1920](https://github.com/opensearch-project/k-NN/pull/1920)
 * Refactor KNNVectorFieldType from KNNVectorFieldMapper to a separate class for better readability. [#1931](https://github.com/opensearch-project/k-NN/pull/1931)
 * Generalize lib interface to return context objects [#1925](https://github.com/opensearch-project/k-NN/pull/1925)
-* Move k search k-NN query to re-write phase of vector search query for Native Engines [#1877](https://github.com/opensearch-project/k-NN/pull/1877)
```
Comment (shatejas) [inline CHANGELOG.md:42]: Removing this, the change is now covered in the log above
```
k-NN query rescore support for native engines [#1984](https://github.com/opensearch-project/k-NN/pull/1984)
```

File: src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java:98]: yeah this looks good

**Review by jmazanec15 — APPROVED** [review]
LGTM

**Review by naveentatikonda — APPROVED** [review]

**Comment by naveentatikonda** [comment]
@shatejas backport to 2.x failed. Can you fix it and raise a PR. Thanks!

**Pull request merged**